src/common/snippets/src/pass/extract_reshapes_from_mha.cpp

```diff
@@ -66,3 +68,104 @@ ov::snippets::pass::ExtractReshapesFromMHA::ExtractReshapesFromMHA() {
+    };
+
+    auto matmul_m = pattern::wrap_type<opset1::MatMul>(static_shape_single_consumer);
+    auto input_1_m = pattern::any_input(pattern::has_static_shape());
+    auto eltwise_1_m = pattern::optional<ov::op::util::BinaryElementwiseArithmetic>({matmul_m, input_1_m},
+                                                                                    static_shape_single_consumer);
+    auto reshape_1_m = pattern::wrap_type<opset1::Reshape>({eltwise_1_m, pattern::wrap_type<opset1::Constant>()},
+                                                           rank_upgrade_reshape);
+    auto input_2_m = pattern::any_input(has_leading_dimension_one);
+    auto eltwise_2_m = pattern::wrap_type<ov::op::util::BinaryElementwiseArithmetic>({reshape_1_m, input_2_m},
+                                                                                     static_shape_single_consumer);
+    auto reshape_2_m = pattern::wrap_type<opset1::Reshape>({eltwise_2_m, pattern::wrap_type<opset1::Constant>()},
+                                                           rank_reduction_reshape);
+
+    matcher_pass_callback callback = [OV_CAPTURE_CPY_AND_THIS](pattern::Matcher& m) {
+        OV_ITT_SCOPED_TASK(ov::pass::itt::domains::SnippetsTransform, "Snippets::op::RankUpgradeToRankReduction")
+        const auto& pattern_map = m.get_pattern_value_map();
+        const auto& matmul = pattern_map.at(matmul_m);
+        const auto matmul_node = ov::as_type_ptr<opset1::MatMul>(matmul.get_node_shared_ptr());
+        if (!ov::snippets::pass::TokenizeMHASnippets::is_matmul0_supported(matmul_node) ||
+            transformation_callback(matmul_node))
+            return false;
+
+        const auto& input_2 = pattern_map.at(input_2_m);
+        auto input_2_shape = input_2.get_shape();
+        input_2_shape.erase(input_2_shape.begin());
+        const auto target_shape = ov::opset1::Constant::create(ov::element::i32, {input_2_shape.size()}, input_2_shape);
+        const auto reshaped_input2 = std::make_shared<ov::opset1::Reshape>(input_2, target_shape, true);
+        const auto& reshape_1 = pattern_map.at(reshape_1_m).get_node_shared_ptr();
+        ov::copy_runtime_info(reshape_1, reshaped_input2);
+
+        auto first_input = pattern_map.at(matmul_m);
+        if (pattern_map.count(eltwise_1_m)) {
+            first_input = pattern_map.at(eltwise_1_m);
+        }
+        const auto& eltwise_2 = pattern_map.at(eltwise_2_m).get_node_shared_ptr();
+        const auto& shapes = ov::util::get_node_input_partial_shapes(*eltwise_2);
+        OPENVINO_ASSERT(shapes.size() > 0, "Eltwise node should has at least one input.");
+        auto equal_rank = [&](const PartialShape& p) {
+            return p.size() == shapes[0].size();
+        };
+        if (!std::all_of(shapes.cbegin(), shapes.cend(), equal_rank)) {
+            return false;
+        }
+
+        const auto new_eltwise_2 = eltwise_2->clone_with_new_inputs({first_input, reshaped_input2});
```

Comment:
By the way, can we just call `set_arguments` in original `Eltwise` op instead of cloning? Then we don't lose rt info, for example.

Reply:
Yeah, it's more light weight. Done, thanks Alexandra!